### PR TITLE
[expo-av] add opt-in metering info to status

### PR DESCRIPTION
**Pull request by danieloi**
# Why

Adds decibel information to the status object. 
This makes it so sound level information can be visualized and/or displayed on android and ios.
# How

- added withMeteringFlag parameter to prepareAudioRecorder function in js and native land
- added necessary logic and API calls for getting metering values in getAudioRecorderStatus function in native land
- return the status object in native land, with or without the metering key, based on the flag value in the prepareAudioRecorder call.
# Test Plan

All tests pass. 
The Recorder example doesn't take in audio in the android emulator but runs as usual with the feature enabled.

### CODE REVIEW

**Review by bbarthec — CHANGES_REQUESTED** [review]
Ok, so I've got a couple of questions:

1. Why in the `results` object you've created:
```ts
{
  ...
  metering: {
    value: number
  }
  ...
}
```
instead of:
```ts
{
  ...
  metering: number
  ...
}
```
I don't see any additional info for this key apart from a single number.

2. Why you've created another parameter `isMeteringEnabled` instead of adding it to the `RecordingOptions`? Change it.

3. As far as I can see when `isMeteringEnabled = true` you're just simply querying the system, so maybe let's abandon this flag all together and add this information always? 🤔 Let's go this way 👍 

4. Why have you used the word `metering`? 🤔 Does it have some specific meaning in regards to decibels/volume/amplitude? 🤔 

Once you give me answers for these questions & adjust the PR according to 2. and 3. I'll resume reviewing.

And there're are parts that are missing:
- CHANGELOG entry
- docs change

**Comment by danieloi** [comment]
> he PR accord

1. That makes sense. I started off with porting this feature from [react-native-sound-level](https://github.com/punarinta/react-native-sound-level) and they used an object with a "value" key and a "raw value" key because android returned a different value from ios. Changed my mind last minute from adding the "raw value" key and should've just used the number instead of an object

2. Because it didn't have to do with ios or android. What do you say I make it a key at the same depth as "ios" and "android"? 

3. [_" Because metering uses computing resources, turn it on only if you intend to use it."_](https://developer.apple.com/documentation/avfoundation/avaudiorecorder/1386355-meteringenabled?language=objc)

4. see No. 3. I borrowed the term from the ios docs. Also noticed [react-native-audio-toolkit](https://github.com/react-native-audio-toolkit/react-native-audio-toolkit/blob/master/docs/API.md) used similar naming.

What are your thoughts  @bbarthec ?

**Comment by bbarthec** [comment]
@danieloi, yeah, the observations of yours seem legit 👍 
Please proceed with this PR work according to your reply 💪

The other thing is that we'll need some explanation in Android codebase (some inline comment) why you're using this piece of code with given constants:
```java
    int amplitude = mAudioRecorder.getMaxAmplitude();
    if (amplitude == 0) {
      result = -160;
    } else {
      result = (int) (20 * Math.log(((double) amplitude) / 32767d));
    }
```

**Comment by danieloi** [comment]
Pardon the delay.

I've made the changes. 

Added comments on the constants per your request. 

I asked to further explain the hex constant (32767d) in the issues of the source repos I mentioned. 
It's supposed to be a ratio between the base sound level and the measured one. Decibels is a relative measurement. 

That value gives levels info comparable to iOS's values.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 1]
```diff
@@ -534,12 +536,46 @@ private long getAudioRecorderDurationMillis() {
...
+      result = -160;
+    } else {
+      /**
+       * amplitude to decibel conversion
+       * 
+       */
+      result = (int) (20 * Math.log(((double) amplitude) / 32767d));
+    }
+
+    return result;
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java:563]:
```suggestion
    int amplitude = mAudioRecorder.getMaxAmplitude();
    if (amplitude == 0) {
      return -160;
    }
     
    // Amplitude to decibel conversion.
    // Code copied from https://github.com/punarinta/react-native-sound-level
    // It's supposed to be a ratio between the base sound level and the measured one.
    // Decibels is a relative measurement.
	// Value `32767d` gives levels info comparable to iOS's values.
	// see: https://github.com/punarinta/react-native-sound-level/issues/20
    return = (int) (20 * Math.log(((double) amplitude) / 32767d));
```

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 2]
```diff
@@ -534,12 +536,46 @@ private long getAudioRecorderDurationMillis() {
...
+      if(mAudioRecorderIsMeteringEnabled) {
+        map.putBoolean("canRecord", true);
+        map.putBoolean("isRecording", mAudioRecorderIsRecording);
+        map.putInt("durationMillis", (int) getAudioRecorderDurationMillis());
+        map.putInt("metering", (int) getAudioRecorderLevels());
+      } else {
+        map.putBoolean("canRecord", true);
+        map.putBoolean("isRecording", mAudioRecorderIsRecording);
+        map.putInt("durationMillis", (int) getAudioRecorderDurationMillis());
+      }
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java:578]:
```suggestion
      map.putBoolean("canRecord", true);
      map.putBoolean("isRecording", mAudioRecorderIsRecording);
      map.putInt("durationMillis", (int) getAudioRecorderDurationMillis());

      if (mAudioRecorderIsMeteringEnabled) {
        map.putInt("metering", (int) getAudioRecorderLevels());
      }
```

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 3]
```diff
@@ -586,6 +622,12 @@ public void prepareAudioRecorder(final ReadableArguments options, final Promise
       return;
     }
 
+    if(options.getBoolean(RECORDING_OPTION_IS_METERING_ENABLED_KEY)) {
+      mAudioRecorderIsMeteringEnabled = true;
+    } else {
+      mAudioRecorderIsMeteringEnabled = false;
+    }
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java:629]:
```suggestion
    mAudioRecorderIsMeteringEnabled
 = options.getBoolean(RECORDING_OPTION_IS_METERING_ENABLED_KEY);
```

File: packages/expo-av/ios/EXAV/EXAV.m [hunk 4]
```diff
@@ -529,6 +530,15 @@ - (NSDictionary *)_getAudioRecorderStatus
...
+        [_audioRecorder updateMeters];
+      float _currentLevel = [_audioRecorder averagePowerForChannel: 0];
+      return @{@"canRecord": @(YES),
+             @"isRecording": @([_audioRecorder isRecording]),
+             @"durationMillis": @(durationMillis),
+             @"metering": @(_currentLevel)};
+    }
     return @{@"canRecord": @(YES),
              @"isRecording": @([_audioRecorder isRecording]),
              @"durationMillis": @(durationMillis)};
```
Comment (bbarthec) [inline packages/expo-av/ios/EXAV/EXAV.m:544]:
```suggestion

    NSMutableDictionary *result = [@{
      @"canRecord": @(YES),
      @"isRecording": @([_audioRecorder isRecording]),
      @"durationMillis": @(durationMillis),
    } mutableCopy];

    if (_audioRecorder.meteringEnabled) {
      [_audioRecorder updateMeters];
      float currentLevel = [_audioRecorder averagePowerForChannel: 0];
      result[@"metering"] = @(currentLevel);
    }

    return result;
```

**Review by bbarthec — COMMENTED** [review]
Overall it looks fine 👍 
I've suggested a few style improvements.

File: packages/expo-av/ios/EXAV/EXAV.m [hunk 5]
```diff
@@ -795,6 +807,9 @@ - (void)sendEventWithName:(NSString *)eventName body:(NSDictionary *)body
       reject(@"E_AUDIO_RECORDERNOTCREATED", [NSString stringWithFormat:@"Prepare encountered an error: %@", error.description], error);
       return;
     } else if ([_audioRecorder prepareToRecord]) {
+        if(options[EXAudioRecordingOptionsIsMeteringEnabledKey]) {
+        _audioRecorder.meteringEnabled = true;
+        }
```
Comment (bbarthec) [inline packages/expo-av/ios/EXAV/EXAV.m:812]:
```suggestion
      if (options[EXAudioRecordingOptionsIsMeteringEnabledKey]) {
        _audioRecorder.meteringEnabled = true;
      }
```

**Review by bbarthec — COMMENTED** [review]
LGTM 👍 
Please, rebase it so there is no conflict anymore and we're ready to go 😉

**Comment by bbarthec** [comment]
@danieloi, I'm not sure what've done here, but it seems that you've pulled in to this PR quite a number of commits that are completely unrelated to the topic 😰

**Comment by danieloi** [comment]
Haha 
I “rebased” to get rid of conflicts. 
I’ll undo it and push again.

**Comment by danieloi** [comment]
I'm getting a lot of baggage when I try to resolve the conflict. @bbarthec. 

It's in the CHANGELOG, it floods my repo with a lot of changes I didn't have before when I click through the resolve flow. 

Help?

**Comment by danieloi** [comment]
Tried adding it to another branch and it looks like there are no more issues.

**Review by bbarthec — APPROVED** [review]
Yeah, looks good now 😊 
Thank you for all the effort you've put into this PR (rebasing, providing context, comments, etc.).
Really great job! 🎉
I hope you'll create some more PRs as you're more experienced with `expo/expo` repository now 😊

**Comment by danieloi** [comment]
Y’all are doing great work. 

Looking for employment at the moment if y’all have anything open for a self starter. 
Shameless plug but there it is 😄. 

Thanks for the timely, valuable feedback and improvements all the way.

**Comment by blainekasten** [comment]
Any chance this can be published in a new version? 😄

**Comment by bbarthec** [comment]
@blainekasten, I've just published a new version of this package to `npm` under the `next` tag. (https://www.npmjs.com/package/expo-av/v/9.0.0). 🎉

**Comment by blainekasten** [comment]
@bbarthec i just installed `next` and it appears to have updated from the js side (type defs are updated, and preset includes the metering flag). But the stats update is not including the `meter` flag. Is there anything special I need to do to make sure the new native code is loaded?

Running on ios through the expo app.

**Comment by bbarthec** [comment]
@blainekasten, this is not going to work in Expo Client until the new SDK containing this change is released 😟

**Pull request merged**